design/src/server/middleware.md

```diff
@@ -0,0 +1,350 @@
+
+Middleware in `tower` typically conforms to the following pattern, a `Service` implementation of the form
+
+```rust
+pub struct NewService<S> {
+    inner: S,
+    /* auxillary data */
+}
+```
+
+and a complementary
+
+```rust
+pub struct NewLayer {
+    /* auxiliary data */
+}
+
+impl<S> Layer<S> for NewLayer {
+    type Service = NewService<S>;
+
+    fn layer(&self, inner: S) -> Self::Service {
+        NewService {
+            inner,
+            /* auxiliary fields */
+        }
+    }
+}
+```
+
+The `NewService` modifies the behavior of the inner `Service` `S` while the `NewLayer` takes auxiliary data and constructs `NewService<S>` from `S`.
+
+Customers are then able to stack middleware by composing `Layer`s using combinators such as [`ServiceBuilder::layer`](https://docs.rs/tower/latest/tower/struct.ServiceBuilder.html#method.layer) and [`Stack`](https://docs.rs/tower/latest/tower/layer/util/struct.Stack.html).
+
+<!-- TODO(Update documentation): There's a `Layer` implementation on tuples about to be merged, give it as an example here. -->
+
+## Applying Middleware
+
+One of the primary goals of Smithy Rust is to provide configurability and extensibility through the application of middleware. The customer is able to apply `Layer`s in a variety of key places during the request/response lifecycle. The following schematic labels each configurable middleware position from A to D:
+
+```mermaid
+stateDiagram-v2
+    state in <<fork>>
+    state "GetPokemonSpecies" as C1
+    state "GetStorage" as C2
+    state "DoNothing" as C3
+    state "..." as C4
+    direction LR
+    [*] --> in : HTTP Request
+    UpgradeLayer --> [*]: HTTP Response
+    state A {
+        state PokemonService {
+            state RoutingService {
+                in --> UpgradeLayer: HTTP Request
+                in --> C2: HTTP Request
+                in --> C3: HTTP Request
+                in --> C4: HTTP Request
+                state B {
+                    state C1 {
+                        state C {
+                            state UpgradeLayer {
+                                direction LR
+                                [*] --> Handler: Model Input
+                                Handler --> [*] : Model Output
+                                state D {
+                                    Handler
+                                }
+                            }
+                        }
+                    }
+                    C2
+                    C3
+                    C4
+                }
+            }
+        }
+    }
+    C2 --> [*]: HTTP Response
+    C3 --> [*]: HTTP Response
+    C4 --> [*]: HTTP Response
+```
+
+where `UpgradeLayer` is the `Layer` converting Smithy model structures to HTTP structures and the `RoutingService` is responsible for routing requests to the appropriate operation.
+
+### A) Outer Middleware
```

Comment:
While I understand it, I believe the difference between A) and B) is not clear in the doc.

Reply:
Hows this?
https://github.com/awslabs/smithy-rs/pull/1844/commits/baf1ca4772bb2837f2c3929f613c27cb8b6beebf